File: ts/simplifyPath.ts
function simplifyPath(path: string): string {
  return (
    '/' +
    path
      .split('/')
      .reduce((result, p) => {
        if (!p.length || p === '.') {
          return result;
        }
        if (p === '..') {
          result.pop();
          return result;
        }

        result.push(p);
        return result;
      }, [] as string[])
      .join('/')
  );
}

export default () => {
  const path = '/a/./b/../../c/';
  // const path = '/.../a/../b/c/../d/./';
  // const path = '/../';
  // const path = '/home//foo/';
  // const path = '/home/';
  // const path = '/home/user/Documents/../Pictures';
  const result = simplifyPath(path);
  console.log(result);
};

File: rs/src/simplify_path.rs
fn simplify_path(path: String) -> String {
    let result: Vec<&str> = vec![];
    return "/".to_owned()
        + &path
            .split('/')
            .fold(result, |mut result, p| match p {
                "" => {
                    return result;
                }
                "." => {
                    return result;
                }
                ".." => {
                    result.pop();
                    return result;
                }
                p => {
                    result.push(p);
                    return result;
                }
            })
            .join("/");
}

pub fn main() {
    let path: String = "/a/./b/../../c/".to_string();
    // let path: String = "/.../a/../b/c/../d/./".to_string();
    let result = simplify_path(path);
    println!("{result}");
}
